pkg/registry/apps/statefulset/strategy.go

```diff
@@ -73,6 +88,19 @@ func (statefulSetStrategy) PrepareForCreate(ctx context.Context, obj runtime.Obj
 	statefulSet.Generation = 1
 
 	pod.DropDisabledTemplateFields(&statefulSet.Spec.Template, nil)
+	dropStatefulSetDisabledFields(statefulSet, nil)
+}
+
+// maxUnavailableInUse returns true if StatefulSet's maxUnavailable set(used)
+func maxUnavailableInUse(statefulset *apps.StatefulSet) bool {
+	if statefulset == nil {
+		return false
+	}
+	if statefulset.Spec.UpdateStrategy.RollingUpdate == nil {
+		return false
+	}
+
+	return statefulset.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable != nil
```

Comment:
Should this require a non-default value as well? Otherwise it's basically a check that the object was written to while the gate was enabled, which doesn't seem to add much value to me.

I'm not sure I see the benefit of feature-gating this in apiserver. The risky stuff is in the controller. https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-unstable-features-to-stable-versions kind of sort of says you can't do this at all?

If we are going to feature gate this, shouldn't we also clear the field on read? But then we change objects in a way that would conflict with copies of those objects in caches.

@liggitt may have thoughts.

Reply:
hmm... other places avoid applying defaults to fields disabled via field-gate:

```
find pkg -name defaults.go | xargs grep '.Enabled('
pkg/apis/core/v1/defaults.go:	if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
pkg/apis/core/v1/defaults.go:	if utilfeature.DefaultFeatureGate.Enabled(features.ServiceLBNodePortControl) {
pkg/apis/scheduling/v1/defaults.go:	if utilfeature.DefaultFeatureGate.Enabled(features.NonPreemptingPriority) && obj.PreemptionPolicy == nil {
pkg/apis/scheduling/v1beta1/defaults.go:	if utilfeature.DefaultFeatureGate.Enabled(features.NonPreemptingPriority) && obj.PreemptionPolicy == nil {
pkg/apis/scheduling/v1alpha1/defaults.go:	if utilfeature.DefaultFeatureGate.Enabled(features.NonPreemptingPriority) && obj.PreemptionPolicy == nil {
pkg/apis/storage/v1/defaults.go:	if obj.Spec.StorageCapacity == nil && utilfeature.DefaultFeatureGate.Enabled(features.CSIStorageCapacity) {
pkg/apis/storage/v1/defaults.go:	if obj.Spec.FSGroupPolicy == nil && utilfeature.DefaultFeatureGate.Enabled(features.CSIVolumeFSGroupPolicy) {
pkg/apis/storage/v1/defaults.go:	if len(obj.Spec.VolumeLifecycleModes) == 0 && utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
pkg/apis/storage/v1/defaults.go:	if obj.Spec.RequiresRepublish == nil && utilfeature.DefaultFeatureGate.Enabled(features.CSIServiceAccountToken) {
pkg/apis/storage/v1beta1/defaults.go:	if obj.Spec.StorageCapacity == nil && utilfeature.DefaultFeatureGate.Enabled(features.CSIStorageCapacity) {
pkg/apis/storage/v1beta1/defaults.go:	if obj.Spec.FSGroupPolicy == nil && utilfeature.DefaultFeatureGate.Enabled(features.CSIVolumeFSGroupPolicy) {
pkg/apis/storage/v1beta1/defaults.go:	if len(obj.Spec.VolumeLifecycleModes) == 0 && utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
pkg/apis/storage/v1beta1/defaults.go:	if obj.Spec.RequiresRepublish == nil && utilfeature.DefaultFeatureGate.Enabled(features.CSIServiceAccountToken) {
pkg/scheduler/apis/config/v1beta1/defaults.go:	if feature.DefaultFeatureGate.Enabled(features.DefaultPodTopologySpread) {
```

I haven't thought through all the implications of that yet...

Reply:
> I'm not sure I see the benefit of feature-gating this in apiserver. The risky stuff is in the controller. https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-unstable-features-to-stable-versions kind of sort of says you can't do this at all?

The main reasons:
* avoid allowing data to be persisted in a new field that will get dropped by a write via a skewed n-1 API server during upgrade
* avoid allowing data into alpha fields when the alpha feature is enabled

Reply:
I think those reasons make sense for gates that have never been on, but once the gate has been on for a while and is turned off, I'm not sure I buy them.

Basically this is my complaint about feature gates: turning them on and then off doesn't get you back to where you started, it gets you into a quasi-state that's less tested than either on or off.

Reply:
i believe no action is required from me ?

Reply:
check the feature gate in defaults.go

Reply:
this is resolved

Reply:
Here's my issue:

T1: gate is off, all objects have a nil value here
T2: gate is on, all objects read a 1 here
T3: gate is off again. Now some objects have a nil, but any object written while the gate was on read a 1 still.

Instead of returning the cluster to a known safe state, we have put the cluster into another even less well tested state.

Reply:
@lavalamp is this an action item for me or general brain storming for a deeper take we need to take in defaulting longer term ?